src/main/java/se/liu/ida/hefquin/engine/queryplan/utils/TextBasedLogicalPlanPrinterImpl2.java

```diff
@@ -48,11 +71,49 @@ public String getIndentLevelString(final int planNumber, final int planLevel, fi
 	 */
 	public void planWalk( final LogicalPlan plan, final int planNumber, final int planLevel, final int numberOfSiblings, final PrintStream out) {
 		final String indentLevelString = getIndentLevelString(planNumber, planLevel, numberOfSiblings);
-		out.append( indentLevelString + plan.getRootOperator().toString() );
-		out.append( System.lineSeparator() );
+		final String indentLevelStringForOpDetail = getIndentLevelStringForDetail(planNumber, planLevel, numberOfSiblings);
+		final LogicalOperator rootOp = plan.getRootOperator();
+		if ( rootOp instanceof LogicalOpMultiwayJoin ) {
+			printOperatorInfoMultiwayJoin( (LogicalOpMultiwayJoin) rootOp, out, indentLevelString );
+		}
+		else if (rootOp instanceof LogicalOpMultiwayUnion) {
+			printOperatorInfoMultiwayUnion( (LogicalOpMultiwayUnion) rootOp, out, indentLevelString );
+		}
+		else if (rootOp instanceof LogicalOpRequest) {
+			printOperatorInfoForRequest( (LogicalOpRequest) rootOp, out, indentLevelString, indentLevelStringForOpDetail );
+		}
+		else {
+			throw new IllegalArgumentException( "Unexpected operator type: " + rootOp.getClass().getName() );
+		}
 		for ( int i = 0; i < plan.numberOfSubPlans(); ++i ) {
 			planWalk( plan.getSubPlan(i), i, planLevel+1, plan.numberOfSubPlans(), out );
 		}
 	}
-
+	
+	protected void printOperatorInfoForRequest( final LogicalOpRequest op, final PrintStream out, final String indentLevelString, final String indentLevelStringForOpDetail ) {
+		final DataRetrievalRequest req = op.getRequest();
+		out.append( indentLevelString + "req (" + op.getID() + ")" );
+		out.append( System.lineSeparator() );
+		printFederationMember( op.getFederationMember(), indentLevelStringForOpDetail, out );
+		out.append( indentLevelStringForOpDetail + "  - pattern (" + req.toString() + ")" );
+		out.append( System.lineSeparator() );
+		out.append( indentLevelStringForOpDetail );
+		out.append( System.lineSeparator() );
+	}
+	
+	protected void printOperatorInfoMultiwayJoin( final LogicalOpMultiwayJoin op, final PrintStream out, final String indentLevelString ) {
+		out.append( indentLevelString + "mj (" + op.getID() + ") " );
+		out.append( System.lineSeparator() );
+	}
+	
+	protected void printOperatorInfoMultiwayUnion( final LogicalOpMultiwayUnion op, final PrintStream out, final String indentLevelString ) {
+		out.append( indentLevelString + "mu (" + op.getID() + ") " );
+		out.append( System.lineSeparator() );
+	}
+	
+	protected void printFederationMember( final FederationMember fm, final String indentLevelStringForOpDetail, final PrintStream out ) {
+		out.append( indentLevelStringForOpDetail + "  - fm (" + fm.getInterface().toString() + ")" );
```

Comment:
Two more things about this code line, which I realize only now:

First, instead of relying on the toString() function of the interface, better use the hashCode() function directly. (In fact, an even better solution would be to introduce unique IDs for interface instances, similar to how you have done for the logical operators during the Malaga event. Yet, I would say that we keep this idea for a follow-up PR, to avoid overloading this PR here.)

Second, you forgot to print the detailed information about the interface, such as
"SPARQL endpoint at https://query.wikidata.org/sparql"

Reply:
As I see, the `toString` function here is to print the interface's information. For instance, as defined in `SPARQLEndpointInterfaceImpl` from `se.liu.ida.hefquin.engine.federation.access.impl.iface`.